### PR TITLE
refactor: improve project code

### PR DESCRIPTION
**Pull request by BlackHole1**
1. reducing context pollution (use `Closures`)
2. add JSDoc(making maintenance easier)
3. Function Encapsulation (closer to a pure function, makes it easier to add unit tests later)
4. Revmoe cache (because there is no effect. The `/notifications` page does not update in real time)
5. Remove `Set`/`Map`(There is no need to use them, ordinary objects are a simple solution)
6. Click on the label to open from a new tab.
7. Fix discussion/alert throw error (We do not support Discussion at this time)

### CODE REVIEW

File: content-script.js [hunk 1]
```diff
@@ -1,157 +1,254 @@
...
+* @param {[number, number, number, number, number, number]} color - Issues/PR label color
+* @returns {[number, number, number, number, number, number]}
+*/
+const parseColor = (() => {
+  return (color) => {
+    const r = +('0x' + color[0] + color[1]);
+    const g = +('0x' + color[2] + color[3]);
+    const b = +('0x' + color[4] + color[5]);
+
+    return [r, g, b, ...rgbToHsl(r, g, b)];
```
Comment (nornagon) [inline content-script.js]: Imo better to name these:
```suggestion
    return {r, g, b, ...rgbToHsl(r, g, b)};
```

File: content-script.js [hunk 2]
```diff
@@ -1,157 +1,254 @@
...
+  return fetch('https://api.github.com/graphql', {
+    headers: {
+      Authorization: `Bearer ${accessToken}`
+    },
+    method: 'POST',
+    body: JSON.stringify({
+      query: gql
+    })
+  })
+  .then(resp => {
```
Comment (nornagon) [inline content-script.js]: use async/await please!

File: content-script.js [hunk 3]
```diff
@@ -1,157 +1,254 @@
...
-  if (!resp.ok) {
-    cache.set(key, null)
-    return null
+/**
+ * query repo and id in /notifications page
+ *
+ * @returns {null | NotificationInfo}
+ */
+const notificationRepoAndID = () => {
+  const notificationsItems = document.querySelectorAll('li.notifications-list-item');
```
Comment (nornagon) [inline content-script.js]: looks like this lost the `:not(.watched)` selector, so we'll re-query all the time?

File: content-script.js [hunk 4]
```diff
@@ -1,157 +1,254 @@
...
-    a.textContent = emojify(label.name)
-    labelContainer.appendChild(a)
+  const result = {};
+
+  for (const item of notificationsItems) {
+    const [ repo, id ] = item.querySelector('div[id^=notification_] > div > p:first-of-type')?.textContent.replace(/\s/g, '').split('#') || [ null, null ];
+
+    if (!repo || !id) {
+      console.warn('parse notification item dom failed');
+     continue;
```
Comment (nornagon) [inline content-script.js]: nit: bad indentation

File: content-script.js [hunk 5]
```diff
@@ -1,157 +1,254 @@
...
-              ${[...issues].map((issue, i) => {
-                return `issue${i}: issueOrPullRequest(number: ${issue}) { ... on Issue { ...IssueFragment } ... on PullRequest { ...PRFragment } }`
-              })}
-            }
-          `
-        })}
-      }
+ 
+/**
+* repository GraphQL
```
Comment (nornagon) [inline content-script.js]:
```suggestion
* Form a GraphQL query for the given notifications
```

File: content-script.js [hunk 6]
```diff
@@ -1,157 +1,254 @@
...
+
+        const info = {
+          id: number,
+          labels: labels.nodes
+        }
+
+        if (typeof result[repo] === 'undefined') {
+          result[repo] = [info];
+        } else {
+      result[repo].push(info);
```
Comment (nornagon) [inline content-script.js]: weird indentation here too

File: content-script.js (hunk 3 above)
Comment (BlackHole1) [reply content-script.js]: We don't need `setInterval` / `MutationObserver` / `cache` and this.

**Reson:**
The github `/notification` page don't have `hot reload`. No changes are made to the front-end UI layer, even if a new notification comes in. Unless we do a page refresh.

Comment (nornagon) [reply content-script.js]: What about when selecting some notifications and clicking 'done'?

Comment (BlackHole1) [reply content-script.js]: oooh. You're right

**Pull request merged**